### PR TITLE
Added pure_bf16 mode

### PR DESCRIPTION
**Pull request by arlesniak**
### PR types

Others
### PR changes

APIs
### Describe

Added pure_bf16 mode for AMP

### CODE REVIEW

**Comment by arlesniak** [comment]
@jczaja @arogowie-intel @wozna Could you start your review, please? PR-CI-APPROVAL, PR-CI-CPU-Py2 fail bof changed API approvals but the core to be reviewed won't be changed.

File: paddle/fluid/operators/amp/fp16_type_traits.h [hunk 1]
```diff
@@ -32,6 +33,12 @@ class MPTypeTrait<platform::float16> {
   using Type = float;
 };
 
+template <>
+class MPTypeTrait<platform::bfloat16> {
+ public:
+  using Type = float;
+};
+
```
Comment (arogowie-intel) [inline paddle/fluid/operators/amp/fp16_type_traits.h]: I think that this file name should be changed to sth like `mp_type_traits.h` cause this is not solely fp16 related file any more.

File: python/paddle/fluid/contrib/mixed_precision/bf16/amp_lists.py [hunk 2]
```diff
@@ -13,8 +13,10 @@
 # limitations under the License.
 
 import copy
+from .... import core
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/amp_lists.py]: IMHO following would be more readable and in case of this file directory change easier to maintain.
```suggestion
from paddle.fluid import core
```

File: paddle/fluid/framework/operator.cc [hunk 3]
```diff
@@ -1061,7 +1061,8 @@ bool OperatorWithKernel::SupportsMKLDNN(
 bool OperatorWithKernel::CanMKLDNNBeUsed(const framework::ExecutionContext& ctx,
                                          proto::VarType::Type data_type) const {
   bool use_mkldnn_ctx =
-      ctx.Attr<bool>("use_mkldnn") && platform::is_cpu_place(ctx.GetPlace());
+      (ctx.Attr<bool>("use_mkldnn") || data_type == proto::VarType::BF16) &&
```
Comment (jczaja) [inline paddle/fluid/framework/operator.cc]: Why this change is needed?

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        parameters by their scaled gradients.
+  
+        Args:
+            params_grads (list): A list of params and scaled grads.
+    
+        Returns:
+            A list of optimize operators.
+        """
+
+        # Change the op_role_var attr for some ops, so that gradients
```
Comment (jczaja) [inline python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py]: Loss scaling is not applicable to bf16. Maybe you can remove commented out code and put some clarifying that this is only for API complience?

File: python/paddle/fluid/tests/book/test_word2vec_book.py [hunk 5]
```diff
@@ -121,14 +128,16 @@ def __network__(words):
 
     def train_loop(main_program):
         exe.run(fluid.default_startup_program())
+        if pure_bf16:
+            sgd_optimizer.amp_init(exe.place)
```
Comment (arogowie-intel) [inline python/paddle/fluid/tests/book/test_word2vec_book.py:132]: I am confused right now. There are two switches `use_bf16` and `pure_bf16`. If one set only `use_bf16` to true then `sgd_optimizer` is "decorated" with `amp.bf16` but it is not `amp_init` .. why?

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py (hunk 4 above)
Comment (arlesniak) [reply python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py]: Right, I'll remove the comment

File: paddle/fluid/framework/operator.cc (hunk 3 above)
Comment (arlesniak) [reply paddle/fluid/framework/operator.cc]: Must have left it during experiments, thanks.

**Review by jczaja — DISMISSED** [review]

File: python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py [hunk 6]
```diff
@@ -191,7 +232,180 @@ def bf16_guard():
         yield
 
 
-def rewrite_program_bf16(main_prog, amp_lists=None, use_bf16_guard=False):
+def cast_model_to_bf16(program, amp_lists=None, use_bf16_guard=True):
+    """
+    Traverse all ops in the whole model and set their inputs and outputs
+    to the bf16 data type. This function will do some special process for
+    the batch normalization, which keeps the computational process of
+    batchnorms in FP32.
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py]:
```suggestion
    to the bf16 data type. This function will do some special processing for
    the batch normalization, which will keep the batchnorm's computations
    in FP32.
```
Additionally why batchnorm is not supported/specially treated? Shouldn't it be in unsupported list then?

File: python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py [hunk 7]
```diff
@@ -191,7 +232,180 @@ def bf16_guard():
...
+        idx = 0
+        while idx < len(ops):
+            op = ops[idx]
+            num_cast_ops = 0
+            if op not in keep_fp32_ops:
+                if op in to_bf16_pre_cast_ops:
+                    in_var_cast_num = _insert_cast_op(block, op, idx,
+                                                      core.VarDesc.VarType.FP32,
+                                                      core.VarDesc.VarType.BF16)
+                    num_cast_ops += in_var_cast_num
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py:347]: This branch would never get into since you're not adding anything to `to_bf16_pre_cast_ops` in above code.

File: python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py [hunk 8]
```diff
@@ -191,7 +232,180 @@ def bf16_guard():
...
+        all_parameters.extend(block.all_parameters())
+
+    bf16_var_names = to_bf16_var_names if to_bf16_var_names else set()
+    var_scope = scope if scope else global_scope()
+    for param in all_parameters:
+        if param.name in bf16_var_names:
+            _logger.debug("---- cast {} to bf16 dtype ----".format(param.name))
+            param_t = var_scope.find_var(param.name).get_tensor()
+            data = np.array(param_t)
+            param_t.set(np.uint16(data), place)
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py]: Do you convert here fp32 do bf16 data? If yes then you should rather use `convert_float_to_uint16`

**Review by arogowie-intel — COMMENTED** [review]
I haven't yet gone through `bf16/decorate.py` file, however have looked at most of files. IMHO the core algorithm functions inside `amp_utils.py` would necessary need refactoring since such long functions with so little comments would be unmaintainable in the future.

File: python/paddle/fluid/tests/book/test_word2vec_book.py (hunk 5 above)
Comment (arlesniak) [reply python/paddle/fluid/tests/book/test_word2vec_book.py:132]: In this very example it's needed when pure mode is True. amp_init has internal conditions on itself and does not do anything when pure mode is False nor test_program param is missing (like here).

File: python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py (hunk 6 above)
Comment (arlesniak) [reply python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py]: BatchNorm op can be run in bf16 if needed (in accuracy/speed experiments ), however the exception in the following code is about X, Z inputs not to be forced to bf16 in pure mode. Other inputs don't have that restriction. Unsupported list is intended to be list of ops not implemented for bf16 datatype (not present in the code), so not being able to be called.

File: python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py (hunk 7 above)
Comment (arlesniak) [reply python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py:347]: Must have left it during experiments, thank you

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+from .... import core
+from .... import default_main_program
+from .... import layers
+from .... import program_guard
+from .... import unique_name
+from .amp_utils import rewrite_program_bf16
+from .amp_utils import cast_model_to_bf16
+from .amp_utils import cast_parameters_to_bf16
+# from .amp_utils import update_role_var_grad
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py]: Please remove this line if it's unnecessary.

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+# limitations under the License.
+
+from .... import core
+from .... import default_main_program
+from .... import layers
+from .... import program_guard
+from .... import unique_name
+from .amp_utils import rewrite_program_bf16
+from .amp_utils import cast_model_to_bf16
+from .amp_utils import cast_parameters_to_bf16
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py]: You can group these imports into one statement.

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .... import core
+from .... import default_main_program
+from .... import layers
+from .... import program_guard
+from .... import unique_name
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py]: You can group these imports into one statement.

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    methods `minimize()`, `backward()`, `apply_gradients()` implemented. 
+    Additionally, it enables the MP training automatically, i.e, the creation 
+    and maintenance of master parameters, scaling of loss, etc.
+
+    Args:
+        optimizer (Optimizer): A common Optimizer object.
+        amp_lists (CustomOpLists): An CustomOpLists object.
+        use_pure_bf16(bool): Whether to use the pure bf16 training. Default False.
+        use_bf16_guard(bool): Whether to use `bf16_guard` when constructing the program.
+                           Default None, which means that its value is equal to `use_pure_bf16`.
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py]: There is inconsistency in this description with the function signature. This function parameter does not have default value.

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+    optimizer, plus the support of mixed-precision pre-training. The object
+    of this class almost has the same behavior as the common optimizer, with the 
+    methods `minimize()`, `backward()`, `apply_gradients()` implemented. 
+    Additionally, it enables the MP training automatically, i.e, the creation 
+    and maintenance of master parameters, scaling of loss, etc.
+
+    Args:
+        optimizer (Optimizer): A common Optimizer object.
+        amp_lists (CustomOpLists): An CustomOpLists object.
+        use_pure_bf16(bool): Whether to use the pure bf16 training. Default False.
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py]: There is inconsistency in this description with the function signature. This function parameter does not have default value.

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+        self._learning_rate = optimizer._learning_rate
+        self._learning_rate_map = optimizer._learning_rate_map
+        self._use_pure_bf16 = use_pure_bf16
+        self._use_bf16_guard = use_bf16_guard
+        self._to_bf16_var_names = None
+
+    def _init_amp_var(self):
+        # Ensure the data type of learning rate vars is float32 (same as the
+        # master parameter dtype)
+        if isinstance(self._optimizer._learning_rate, float):
```
Comment (arogowie-intel) [inline python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py:58]: What happens if it is not of `float` data type?

File: python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py [hunk 15]
```diff
@@ -191,7 +232,180 @@ def bf16_guard():
...
+                    if out_var is None or out_var.type not in _valid_types:
+                        continue
+
+                    if out_var.dtype == core.VarDesc.VarType.FP32:
+                        out_var.desc.set_dtype(core.VarDesc.VarType.BF16)
+
+                    _logger.debug(
+                        "-- op type: {}, out var name: {}, out var dtype: {} --".
+                        format(op.type, out_var_name, out_var.dtype))
+            if op.has_attr('in_dtype') and op.attr(
```
Comment (wozna) [inline python/paddle/fluid/contrib/mixed_precision/bf16/amp_utils.py]: Maybe you can put all this dtype attrs into one loop: 
```suggestion
            for attr_name in ['in_dtype', 'out_dtype', 'dtype']:
                  if op.has_attr(attr_name) and op.attr(
                           attr_name) == core.VarDesc.VarType.FP32:
                       op._set_attr(attr_name, core.VarDesc.VarType.BF16)
```

File: python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py (hunk 14 above)
Comment (arlesniak) [reply python/paddle/fluid/contrib/mixed_precision/bf16/decorator.py:58]: During update learning rate multiplied by gradient can be small enough to exceed dynamic range of bf16.

**Comment by arlesniak** [comment]
@wozna @jczaja @arogowie-intel please continue your review

**Review by wozna — APPROVED** [review]
LGTM

**Review by arogowie-intel — APPROVED** [review]

**Comment by arlesniak** [comment]
@luotao1 Could you please start your review ?

**Review by luotao1 — APPROVED** [review]
LGTM

**Comment by luotao1** [comment]
Please cherry-pick into release/2.1
